File: Models/DogBreed.cs
using CommunityToolkit.Mvvm.ComponentModel;

namespace BreedFinder.Models;

public partial class DogBreed : ObservableObject
{
    public string Name { get; set; } = string.Empty;
    public string Description { get; set; } = string.Empty;
    public string Origin { get; set; } = string.Empty;
    public string Group { get; set; } = string.Empty;
    public string Emoji { get; set; } = "🐕";
    public string AccentColor { get; set; } = "#6366F1";
    public string DogCeoBreedKey { get; set; } = string.Empty;

    // Traits (0–10 scale)
    public double Size { get; set; }
    public double EnergyLevel { get; set; }
    public double Shedding { get; set; }
    public double Guardedness { get; set; }
    public double Aggressiveness { get; set; }
    public double Immunity { get; set; }
    public double Lifespan { get; set; }
    public double GroomingNeeds { get; set; }
    public double VetVisitsRequired { get; set; }
    public double Loyalty { get; set; }
    public double SingleOwner { get; set; }

    // Computed at match time
    public double MatchScore { get; set; }
    public string MatchPercentText => $"{(int)(MatchScore * 100)}%";

    // Image — observable so the card updates when the URL arrives from the API
    [ObservableProperty]
    [NotifyPropertyChangedFor(nameof(HasImage))]
    private string imageUrl = string.Empty;

    public bool HasImage => !string.IsNullOrEmpty(ImageUrl);

    public List<string> Tags { get; set; } = [];

    // Shows up to 2 tags as inline text on the compact grid card
    public string TagsDisplay => Tags.Count > 0
        ? string.Join("  ·  ", Tags.Take(2))
        : string.Empty;

    public void ComputeTags()
    {
        Tags.Clear();
        if (Aggressiveness <= 3 && EnergyLevel >= 5) Tags.Add("Family Friendly");
        if (Guardedness >= 7) Tags.Add("Guard Dog");
        if (EnergyLevel >= 8) Tags.Add("Very Active");
        else if (EnergyLevel <= 3) Tags.Add("Low Energy");
        if (Shedding <= 2) Tags.Add("Hypoallergenic");
        if (GroomingNeeds <= 2 && VetVisitsRequired <= 3) Tags.Add("Low Maintenance");
        if (Size <= 3) Tags.Add("Apartment Friendly");
        if (Aggressiveness <= 2 && Immunity >= 7) Tags.Add("Great for Beginners");
        if (Loyalty >= 9) Tags.Add("Highly Loyal");
        if (SingleOwner >= 7) Tags.Add("One-Person Dog");
        if (Lifespan >= 9) Tags.Add("Long Lifespan");
    }
}

File: BreedFinder/Models/DogBreed.swift
import Foundation

// MARK: - Dog Breed model
/// Value-type model that carries all static breed data plus mutable
/// match-time state (score and fetched image URL).
struct DogBreed: Identifiable, Equatable {

    // ── Identity ──────────────────────────────────────────────────────────────
    let id: UUID

    // ── Static data ───────────────────────────────────────────────────────────
    let name: String
    let description: String
    let origin: String
    let group: String
    let emoji: String
    /// Hex string e.g. "#E8892B" – used as the breed's accent tint
    let accentHex: String
    /// Path component(s) for the Dog CEO API, e.g. "retriever/golden"
    let dogCeoBreedKey: String

    // ── Core traits (0–10 scale) ──────────────────────────────────────────────
    let size: Double
    let energyLevel: Double
    let shedding: Double
    let guardedness: Double
    let aggressiveness: Double
    let immunity: Double
    let lifespan: Double
    let groomingNeeds: Double
    let vetVisitsRequired: Double
    let loyalty: Double
    let singleOwner: Double

    // ── Human-centric traits (0–10 scale) ────────────────────────────────────
    /// How good with children (10 = perfect family dog)
    let kidsCompatibility: Double
    /// How well it gets along with other pets (10 = loves all animals)
    let petCompatibility: Double
    /// How well it handles being home alone (10 = very independent)
    let aloneTolerance: Double
    /// How much it barks (10 = very vocal)
    let barkingLevel: Double
    /// Ease of training (10 = learns in a heartbeat)
    let trainability: Double

    // ── Match-time state (mutated by BreedDatabase.getMatchingBreeds) ─────────
    var matchScore: Double = 0
    var imageUrl: String? = nil

    // ── Computed helpers ──────────────────────────────────────────────────────
    var matchPercent: Int { Int(matchScore * 100) }
    var hasImage: Bool { !(imageUrl?.isEmpty ?? true) }

    /// Tags derived purely from trait values — no mutable state needed.
    var tags: [String] {
        var result: [String] = []
        if kidsCompatibility >= 9                           { result.append("Great with Kids") }
        if petCompatibility >= 8                            { result.append("Good with Pets") }
        if trainability >= 9                                { result.append("Easy to Train") }
        if barkingLevel <= 3                                { result.append("Quiet Dog") }
        if aloneTolerance >= 7                              { result.append("Can Be Left Alone") }
        if guardedness >= 7                                 { result.append("Guard Dog") }
        if energyLevel >= 8                                 { result.append("Very Active") }
        else if energyLevel <= 3                            { result.append("Low Energy") }
        if shedding <= 2                                    { result.append("Hypoallergenic") }
        if groomingNeeds <= 2 && vetVisitsRequired <= 3    { result.append("Low Maintenance") }
        if size <= 3                                        { result.append("Apartment Friendly") }
        if loyalty >= 9                                     { result.append("Highly Loyal") }
        if lifespan >= 9                                    { result.append("Long Lifespan") }
        return result
    }

    /// First two tags joined by " · " — shown on compact grid cards.
    var tagsDisplay: String {
        Array(tags.prefix(2)).joined(separator: "  ·  ")
    }

    static func == (lhs: DogBreed, rhs: DogBreed) -> Bool { lhs.id == rhs.id }

    // ── Convenience initialiser (auto-generates id) ───────────────────────────
    init(
        name: String, description: String, origin: String, group: String,
        emoji: String, accentHex: String, dogCeoBreedKey: String,
        size: Double, energyLevel: Double, shedding: Double,
        guardedness: Double, aggressiveness: Double, immunity: Double,
        lifespan: Double, groomingNeeds: Double, vetVisitsRequired: Double,
        loyalty: Double, singleOwner: Double,
        kidsCompatibility: Double, petCompatibility: Double,
        aloneTolerance: Double, barkingLevel: Double, trainability: Double,
        imageUrl: String? = nil
    ) {
        self.id = UUID()
        self.name = name; self.description = description
        self.origin = origin; self.group = group
        self.emoji = emoji; self.accentHex = accentHex
        self.dogCeoBreedKey = dogCeoBreedKey
        self.size = size; self.energyLevel = energyLevel
        self.shedding = shedding; self.guardedness = guardedness
        self.aggressiveness = aggressiveness; self.immunity = immunity
        self.lifespan = lifespan; self.groomingNeeds = groomingNeeds
        self.vetVisitsRequired = vetVisitsRequired
        self.loyalty = loyalty; self.singleOwner = singleOwner
        self.kidsCompatibility = kidsCompatibility
        self.petCompatibility = petCompatibility
        self.aloneTolerance = aloneTolerance
        self.barkingLevel = barkingLevel
        self.trainability = trainability
        self.imageUrl = imageUrl
    }
}
